src/ORM/Behavior/Translate/ShadowTableStrategy.php

```diff
@@ -0,0 +1,591 @@
+            }
+
+            if (in_array($field, $fields)) {
+                $joinRequired = true;
+                $expression->setField("$alias.$field");
+
+                return;
+            }
+
+            if (in_array($field, $mainTableFields)) {
+                $expression->setField("$mainTableAlias.$field");
+            }
+        });
+
+        return $joinRequired;
+    }
+
+    /**
+     * Modifies the entity before it is saved so that translated fields are persisted
+     * in the database too.
+     *
+     * @param \Cake\Event\Event $event The beforeSave event that was fired
+     * @param \Cake\Datasource\EntityInterface $entity The entity that is going to be saved
+     * @param \ArrayObject $options the options passed to the save method
+     * @return void
+     */
+    public function beforeSave(Event $event, EntityInterface $entity, ArrayObject $options): void
+    {
+        $locale = $entity->get('_locale') ?: $this->getLocale();
+        $newOptions = [$this->translationTable->getAlias() => ['validate' => false]];
+        $options['associated'] = $newOptions + $options['associated'];
+
+        // Check early if empty translations are present in the entity.
+        // If this is the case, unset them to prevent persistence.
+        // This only applies if $this->_config['allowEmptyTranslations'] is false
+        if ($this->_config['allowEmptyTranslations'] === false) {
+            $this->unsetEmptyFields($entity);
+        }
+
+        $this->bundleTranslatedFields($entity);
+        $bundled = $entity->get('_i18n') ?: [];
+        $noBundled = count($bundled) === 0;
+
+        // No additional translation records need to be saved,
+        // as the entity is in the default locale.
+        if ($noBundled && $locale === $this->getConfig('defaultLocale')) {
+            return;
+        }
+
+        $values = $entity->extract($this->translatedFields(), true);
+        $fields = array_keys($values);
+        $noFields = empty($fields);
+
+        // If there are no fields and no bundled translations, or both fields
+        // in the default locale and bundled translations we can
+        // skip the remaining logic as its not necessary.
+        if ($noFields && $noBundled || ($fields && $bundled)) {
+            return;
+        }
+
+        $primaryKey = (array)$this->table->getPrimaryKey();
+        $id = $entity->get(current($primaryKey));
+
+        // When we have no key and bundled translations, we
+        // need to mark the entity dirty so the root
+        // entity persists.
+        if ($noFields && $bundled && !$id) {
+            foreach ($this->translatedFields() as $field) {
+                $entity->setDirty($field, true);
+            }
+
+            return;
+        }
+
+        if ($noFields) {
+            return;
+        }
+
+        $where = compact('id', 'locale');
+
+        $translation = $this->translationTable->find()
+            ->select(array_merge(['id', 'locale'], $fields))
```

Comment:
I think for safety sake, these fields should be aliased

Reply:
The fields in `$fields` aren't aliased either, plus the translation table doesn't have any associations / containments so there's no possibility of duplicate field name errors.